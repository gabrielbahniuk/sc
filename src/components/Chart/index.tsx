import React, { useRef, useEffect } from 'react'
import { Chart as ChartJs } from 'chart.js'
import useTimeSeries from '@/hooks/useTimeSeries'
import { appConfig } from '@/config'

type ConeProps = {
  riskLevel: number
  mu: number
  sigma: number
}

type ChartProps = {
  yearsInvestment: number
  cones: ConeProps[]
  investmentValue: number
  riskLevel: number
}

const Chart: React.FC<ChartProps> = ({ investmentValue, yearsInvestment, riskLevel, cones }) => {
  const canvas = useRef<HTMLCanvasElement>(null)

  const { mu, sigma } = cones.filter((cone) => cone.riskLevel === riskLevel)[0]
  const { FEE, MONTHLY_SUM } = appConfig
  ChartJs.defaults.global.defaultFontColor = '#fff'

  const { chartData } = useTimeSeries({
    years: yearsInvestment,
    initialSum: investmentValue,
    fee: FEE,
    mu,
    sigma,
    monthlySum: MONTHLY_SUM
  })

  useEffect(() => {
    drawChart()
  }, [])

  function drawChart(): void {
    const { labels, dataBad, dataGood, dataMedian } = chartData
    const data = {
      datasets: [
        {
          data: dataGood,
          label: 'Good performance',
          borderColor: 'rgba(100, 255, 100, 0.4)',
          fill: false,
          pointRadius: 1
        },
        {
          data: dataMedian,
          label: 'Median performance',
          borderColor: 'rgba(100, 100, 100, 0.4)',
          fill: false,
          pointRadius: 1
        },
        {
          data: dataBad,
          label: 'Bad performance',
          borderColor: 'rgba(255, 100, 100, 0.4)',
          fill: false,
          pointRadius: 1
        }
      ],
      labels
    }

    const options = {
      responsive: false,
      scales: {
        xAxes: [
          {
            display: true,
            scaleLabel: {
              display: true,
              labelString: 'Years',
              fontSize: 16
            },
            gridLines: {
              drawOnChartArea: false
            }
          }
        ],
        yAxes: [
          {
            display: true,
            scaleLabel: {
              display: true,
              fontSize: 16,
              labelString: 'Valuation (EUR)'
            }
          }
        ]
      }
    }

    const config = {
      type: 'line',
      data,
      options
    }

    const context = canvas.current?.getContext('2d')
    new ChartJs(context, config)
  }

  return <canvas ref={canvas} />
}

export default Chart
