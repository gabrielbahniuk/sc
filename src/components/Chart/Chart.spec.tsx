import React from 'react'
import { create } from 'react-test-renderer'
import Chart from '.'

describe('Test Chart component', () => {
  it('should render Chart', () => {
    const props = {
      investmentValue: 10,
      riskLevel: 3,
      yearsInvestment: 1,
      cones: [{ riskLevel: 3, mu: 0.0216, sigma: 0.0215 }]
    }
    const chart = create(<Chart {...props} />).toJSON()
    expect(chart).toMatchSnapshot()
  })
})
