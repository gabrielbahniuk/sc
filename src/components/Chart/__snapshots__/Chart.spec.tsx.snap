// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Chart component should render Chart 1`] = `<canvas />`
